controllers/ingresscluster/ingress-cluster-controller.go

```diff
@@ -0,0 +1,294 @@
+	"github.com/kubevirt/hyperconverged-cluster-operator/controllers/reqresolver"
+	"github.com/kubevirt/hyperconverged-cluster-operator/pkg/downloadhost"
+	hcoutil "github.com/kubevirt/hyperconverged-cluster-operator/pkg/util"
+)
+
+const (
+	ingressName     = "cluster"
+	componentName   = "virt-downloads"
+	secretNamespace = "openshift-config"
+)
+
+var (
+	selfNamespace = hcoutil.GetOperatorNamespaceFromEnv()
+	logger        = logf.Log.WithName("controller_ingress_cluster")
+)
+
+type ReconcileIngressCluster struct {
+	client.Client
+	// used to trigger events in the hyperconverged-controller
+	ingressEventCh chan<- event.TypedGenericEvent[client.Object]
+}
+
+func newIngressClusterController(cl client.Client, ingressEventCh chan<- event.TypedGenericEvent[client.Object]) *ReconcileIngressCluster {
+	return &ReconcileIngressCluster{
+		Client:         cl,
+		ingressEventCh: ingressEventCh,
+	}
+}
+
+func RegisterReconciler(mgr manager.Manager, ingressEventCh chan<- event.TypedGenericEvent[client.Object]) error {
+	return add(mgr, newIngressClusterController(mgr.GetClient(), ingressEventCh))
+}
+
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	c, err := controller.New("ingress-cluster", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	return c.Watch(source.Kind(mgr.GetCache(), client.Object(&configv1.Ingress{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: ingressName,
+		},
+	}),
+		&operatorhandler.InstrumentedEnqueueRequestForObject[client.Object]{},
+		predicate.ResourceVersionChangedPredicate{}),
+	)
+}
+
+func (r *ReconcileIngressCluster) Reconcile(ctx context.Context, _ reconcile.Request) (reconcile.Result, error) {
+
+	clusterIngress := &configv1.Ingress{}
+
+	if err := r.Get(ctx, client.ObjectKey{Name: ingressName}, clusterIngress); err != nil {
+		if errors.IsNotFound(err) {
+			logger.Info("Ingress resource not found. Ignoring since object must be deleted")
+			return reconcile.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	hcExists, err := r.hyperConvergedExists(ctx, logger)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	dh, dhErr := r.getDownloadHost(ctx, clusterIngress)
+
+	if downloadhost.Set(dh) {
+		logger.Info("Notifying the hyperconverged controller about the new hostname")
+		r.ingressEventCh <- event.TypedGenericEvent[client.Object]{Object: clusterIngress}
+	}
+
+	routeInd, needUpdate := updateComponentInStatus(clusterIngress, hcExists, dh)
+
+	if needUpdate {
+		if dhErr != nil {
+			meta.SetStatusCondition(&clusterIngress.Status.ComponentRoutes[routeInd].Conditions, metav1.Condition{
+				Type:    "Degraded",
+				Status:  metav1.ConditionTrue,
+				Reason:  "WrongConfiguration",
+				Message: dhErr.Error(),
+			})
+		} else {
+			meta.SetStatusCondition(&clusterIngress.Status.ComponentRoutes[routeInd].Conditions, metav1.Condition{
+				Type:   "Degraded",
+				Status: metav1.ConditionFalse,
+				Reason: "AsExpected",
+			})
```

Comment:
I think `message` is a required field in a condition.
If the final result will be an empty string in the message field, i think you should put "All is well".